### PR TITLE
Test mempool grpc problem

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

This PR is just to show a problem i am having, is dirty test code and should not be merged.

Most of the changes just add debugging, the real thing we want to test is at `send_transaction_test.rs`. This test runs with an pretty in sync zebra cache and lightwalletd database as follows:

```
cargo test sending_transactions_using_lightwalletd --features lightwalletd-grpc-tests -- --ignored --nocapture 
```

The problem is transactions never get into the mempool however from the logs i can see transactions being added to the storage.

The full log is at https://gist.github.com/oxarbitrage/05d886ef06f77f64898e63dade580899

From this i can see transactions are inserted to the mempool at some point:

```
May 25 18:51:22.538  INFO {zebrad="bae470b" net="Main"}:{peer=Out("8.209.80.185:8233")}:msg_as_req{msg="inv"}: zebrad::components::mempool::storage: storage: Transaction inserted to the mempool: Some(Legacy(transaction::Hash("12f2c78014381178f5d9ca49e4012210af4c34c55e6246de9ea6e3954431d246")))
May 25 18:51:23.098  INFO {zebrad="bae470b" net="Main"}:{peer=Out("135.181.18.180:8233")}:msg_as_req{msg="inv"}: zebrad::components::mempool::storage: storage: Transaction inserted to the mempool: Some(Legacy(transaction::Hash("0b42fb00c056df24dc7bfd5d732afa8a44a012851a475cf91dc37ab71b29eedd")))
```

However this are not the same that are added by the test:

```
May 25 18:52:08.153  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 01b2f63850bb9761eda536f21675611d4dd665ed6fc5dd8a972563e0ee3e3a4a
May 25 18:52:08.234  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 41cfe2a9f8e50da1dbc4588d49501b0dddf72495630858d08d8e29045c0ed1d4
May 25 18:52:08.236  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 08094a565a342248272c7f77241e93f43eb93ae10ee7747eabbff14b8acf4683
May 25 18:52:08.238  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 0e896a3f0ba90ebf69181ddf6c4aa51296c88e3beae9d5d9816b97f3531d5f24
May 25 18:52:08.243  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: febb23fc628c9f3ceb2f756f5078a89209da5997a974141b7ce944a0f1f82763
May 25 18:52:08.247  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 3c8b15452cdc7f9f1ce0984a9f35aae3762973102836e5faa2f89249bc0ef261
May 25 18:52:08.249  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 1ef74dd4c4f7eed44507cd816699076f3a072a14813943ed54b733587ac5b97a
May 25 18:52:08.252  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: a6f6f050dbba321d1e6e8a643814c223dbb505f2078f4b70875829ffd903103f
May 25 18:52:08.255  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: c9ebbd903563cc4029db7c2fd7bfef38610e804114c9143957e2b594d8c97a41
May 25 18:52:08.257  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 15b60f6530437bfedf9816e11b9376b01abc436aa6cf6a1b9399dba65407f33d
May 25 18:52:08.260  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 7326a4a9e766f051142a73e9c9c713815382cd6b15503dab50c02a7321b65b49
May 25 18:52:08.262  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 408183159878d80bc7b4163475a47af928cfcd8cf6ef3d896e777dd39710a4c1
May 25 18:52:08.265  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 8e052f33c4894074808ffaf988bed39ef3feac36f37cfaf66dbe41d5ac022b35
May 25 18:52:08.267  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 7b975f66c0181bcbb43e57fa26e0d482a93ebab7ad43b7d88589429b5cd33008
May 25 18:52:08.270  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: d99d9fca5b717cf3fd82872457b82d5a654a7eef4e88d5fb495d2af1f67a1f9d
May 25 18:52:08.272  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: d29ca96f07af457be7ae232ccfb330605992e97a35484e1ebba4c6320fbc504a
May 25 18:52:08.275  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: f72b41783f6a623f44c207def0c3b66139fd1eb385eff818f726a9b28b1dc656
May 25 18:52:08.278  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 804a3c7c3c950a43c5eedc4e6436d81838cb9ae9265343b33dd6f88678b05a93
May 25 18:52:08.280  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: c737677b988ff86a96b5bf4a4536f8b90e117bb50fde44c7772a2c6a9ee6c5b2
May 25 18:52:08.283  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 0b08364aac3498d05a65411d1e4629c15f1504d1348651054e41cc228d437f43
May 25 18:52:08.285  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 1f571165162f76756d1d20d4d15c0d2085115ff22df298a4512149d016b683fe
May 25 18:52:08.288  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 15a48428e66c70643b9ae0096ed9ebf776bd8ba2fa377538e907f7f7488d3004
May 25 18:52:08.291  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 59b985c8df8f8e7402fd01de06a474eb3de8c9a0e33f4d32715503397a97860b
May 25 18:52:08.293  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: fe4b3c9e9d42f258bb346cedb6759a1672a4b7979154f6977f667b7dd0cd8a66
May 25 18:52:08.296  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: a3da44c925690615b11145f4d4252f33e060b9b2219387c9af39338ae5c609d3
May 25 18:52:08.298  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 82c96af0f9726510a35edc51e9bf40604db0fb90bd7c652d49a235d9035c7401
May 25 18:52:08.301  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 87868cba598062d2f39d3b50c8294686fabdeacd4449cdb6c7764e8033e24331
May 25 18:52:08.303  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: b16673a657575f693267de95c9087b062ce1ff3b37ebd41d30e95351da3a83e8
May 25 18:52:08.309  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: c111e063e3d9d7e4f65f06c562827deb3a81aa20a9958396e0d1642cee2832a8
May 25 18:52:08.315  INFO zebra_rpc::methods: send_raw_transaction: Transaction was sent to the mempool: 8f4f46eb1b68fe7ab3f4d213d9156c3a22e44412eae2e37118e359286434d209
```
When the test `get_mempool_tx` lightwalletd will be calling `get_raw_mempool` however the mempool is always empty:

```
May 25 18:53:17.795  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 18:53:22.794  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 18:53:27.795  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 18:53:32.796  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 18:53:37.793  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
```

### CODE REVIEW

File: zebrad/tests/common/config.rs [hunk 1]
```diff
@@ -77,6 +77,6 @@ pub fn persistent_test_config() -> Result<ZebradConfig> {
 pub fn testdir() -> Result<TempDir> {
     tempfile::Builder::new()
         .prefix("zebrad_tests")
-        .tempdir()
+        .tempdir_in("/media/oxarbitrage/4eb53770-4e4f-4d80-a830-0914f3f5b89a/zebra/issue4350")
```
Comment (oxarbitrage) [inline zebrad/tests/common/config.rs:80]: TODO: create a ticket to make this argument an env variable. The `send_transaction_test` copies the full blockchain to this directory so it is very possible that the user running the test to be out of space in its "hda" drive.

**Comment by oxarbitrage** [comment]
So, i also tried with a custom dummy transaction in the `lightwalletd_wallet_grpc_tests` test:

```
cargo test lightwalletd_wallet_grpc_tests --features lightwalletd-grpc-tests -- --ignored --nocapture
```

Full log is at https://gist.github.com/oxarbitrage/bc4a9932c37fd2ededb6c3b7ae1ccb82

From here i can see the transaction going into the rpc queue:

```
May 25 20:18:32.262  INFO {zebrad="0f879b7" net="Main"}: zebra_rpc::queue: rpc_queue:retry: Attempt to send transaction to mempool: Legacy(transaction::Hash("5ada6b87c22f9fdb424663d79c31d712225d5256445ddd36e179e0dcf550c680"))
```

But never actually getting into the mempool either: 

```
May 25 20:18:34.455  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 20:18:39.456  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
May 25 20:18:44.457  INFO zebra_rpc::methods: get_raw_mempool: Transactions in the mempool: []
```

**Comment by oxarbitrage** [comment]
I added more logs in https://github.com/ZcashFoundation/zebra/pull/4500/commits/8c613a34ffd39435d53c2776d7f1e3628dfb4381

https://gist.github.com/oxarbitrage/9a88a1f67194e369ff42b9e866fb9fe8 for sending transaction test where transactions are from a future block of the blockchain and https://gist.github.com/oxarbitrage/9f12fc50926d785a24ae1c3f85899ec2 for a dummy transaction attempt.

For the sending transaction tests where transactions from a future block are used, they fail at https://github.com/ZcashFoundation/zebra/blob/8c613a34ffd39435d53c2776d7f1e3628dfb4381/zebrad/src/components/mempool.rs#L347-L353

**Comment by oxarbitrage** [comment]
The errors are that the transactions are already in the state:
```
[zebrad/src/components/mempool.rs:348] &e.to_string() = "transaction is already in state"
[zebrad/src/components/mempool.rs:353] txid = Legacy(
    transaction::Hash(
        "8f4f46eb1b68fe7ab3f4d213d9156c3a22e44412eae2e37118e359286434d209",
    ),
)
```

**Comment by teor2345** [comment]
> The errors are that the transactions are already in the state:
> 
> ```
> [zebrad/src/components/mempool.rs:348] &e.to_string() = "transaction is already in state"
> [zebrad/src/components/mempool.rs:353] txid = Legacy(
>     transaction::Hash(
>         "8f4f46eb1b68fe7ab3f4d213d9156c3a22e44412eae2e37118e359286434d209",
>     ),
> )
> ```

That's weird, I thought the test was meant to:
1. advance Zebra to the tip
2. Get some transactions
3. Restart Zebra with a rolled back tip
4. Re-submit those transactions

Maybe the test isn't quite working the way it's supposed to.

**Comment by oxarbitrage** [comment]
In the other test where i try to send a dummy transaction the error seems to be:

```
May 26 16:23:27.768  INFO {zebrad="0f879b7" net="Main"}: zebra_rpc::queue: rpc_queue:retry: Attempt to send transaction to mempool: Legacy(transaction::Hash("5ada6b87c22f9fdb424663d79c31d712225d5256445ddd36e179e0dcf550c680"))
May 26 16:23:27.770 DEBUG {zebrad="0f879b7" net="Main"}:mempool:download_if_needed_and_verify{txid=5ada6b87c22f9fdb424663d79c31d712225d5256445ddd36e179e0dcf550c680}: zebrad::components::mempool::downloads: transaction id already queued for inbound download: ignored transaction txid=Legacy(transaction::Hash("5ada6b87c22f9fdb424663d79c31d712225d5256445ddd36e179e0dcf550c680")) queue_len=1 MAX_INBOUND_CONCURRENCY=10
```

**Comment by oxarbitrage** [comment]
so, the dummy transaction look like this:

```
[zebrad/tests/common/lightwalletd/wallet_grpc_test.rs:266] &transaction = V1 {
    inputs: [
        PrevOut {
            outpoint: OutPoint {
                hash: transaction::Hash(
                    "a049203ce720ba6f3e3d5fbb9fe500fe7b804306e6d2233ab5c8e0331fd82199",
                ),
                index: 0,
            },
            unlock_script: Script(
                "",
            ),
            sequence: 4294967295,
        },
    ],
    outputs: [
        Output {
            value: Amount<zebra_chain::amount::NonNegative>(
                1000000,
            ),
            lock_script: Script(
                "76a914588cff9d4339d754758ade214b3edc69ce57b7f588ac",
            ),
        },
    ],
    lock_time: Height(
        Height(
            0,
        ),
    ),
}
```

It seems that might be too old (V1), i might try to do this with a V4 transaction and see if that gets added.

**Comment by oxarbitrage** [comment]
I need to replace https://github.com/ZcashFoundation/zebra/blob/main/zebra-test/src/vectors.rs#L154 with something more up to date that will maybe get into the mempool.

<strike>@conradoplg  i know you created some transactions, instead of sending them to the blockchain will you be able to get the generated hex of it so i can try to see if it gets into the mempool. Maybe @teor2345 can do this as well. If none can i will try to do it myself but i am trying to save some time as i am around this test since too much already.</strike> Nevermind, i am working on this now.

**Comment by conrado** [comment]
@oxarbitrage I think you meant @conradoplg maybe?

**Comment by oxarbitrage** [comment]
> Maybe the test isn't quite working the way it's supposed to.

It seems it is not really pushing the transactions into the mempool. But we can try the other option which is just send a dummy transaction that gets accepted by the mempool from `lightwalletd_wallet_grpc_tests` so we don't have to deal with `sending_transactions_using_lightwalletd`

**Comment by oxarbitrage** [comment]
> @oxarbitrage I think you meant @conradoplg maybe?

oops, sorry, i fixed that. thanks!

**Comment by oxarbitrage** [comment]
Replaced by https://github.com/ZcashFoundation/zebra/pull/4537

**Pull request closed without merging**